src/txrequest.cpp

```diff
@@ -0,0 +1,238 @@
+        if (priority_new < priority_old) {
+            // There is a CANDIDATE_BEST entry already, but this one is better.
+            Modify<ByTxid>(std::prev(it), [](Entry& entry){ entry.SetState(State::CANDIDATE_READY); });
+            Modify<ByTxid>(it, [](Entry& entry){ entry.SetState(State::CANDIDATE_BEST); });
+        }
+    }
+}
+
+void TxRequestTracker::ChangeAndReselect(typename TxRequestTracker::Index::index<ByTxid>::type::iterator it,
+    TxRequestTracker::State new_state)
+{
+    if (it->IsSelected()) {
+        auto it_next = std::next(it);
+        // The next best CANDIDATE_READY, if any, immediately follows the REQUESTED or CANDIDATE_BEST entry in the
+        // ByTxid index.
+        if (it_next != m_index.get<ByTxid>().end() && it_next->m_txid == it->m_txid &&
+            it_next->GetState() == State::CANDIDATE_READY) {
+            // If one such CANDIDATE_READY exists (for this txid), convert it to CANDIDATE_BEST.
+            Modify<ByTxid>(it_next, [](Entry& entry){ entry.SetState(State::CANDIDATE_BEST); });
+        }
+    }
+    Modify<ByTxid>(it, [new_state](Entry& entry){ entry.SetState(new_state); });
+    assert(!it->IsSelected());
+}
+
+bool TxRequestTracker::MakeCompleted(typename TxRequestTracker::Index::index<ByTxid>::type::iterator it)
+{
+    // Nothing to be done if it's already COMPLETED.
+    if (it->GetState() == State::COMPLETED) return true;
+
+    if ((it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != it->m_txid) &&
+        (std::next(it) == m_index.get<ByTxid>().end() || std::next(it)->m_txid != it->m_txid ||
+        std::next(it)->GetState() == State::COMPLETED)) {
+        // This is the first entry for this txid, and the last non-COMPLETED one. There are only COMPLETED ones left.
+        // Delete them all.
+        uint256 txid = it->m_txid;
+        do {
+            it = Erase<ByTxid>(it);
+        } while (it != m_index.get<ByTxid>().end() && it->m_txid == txid);
+        return false;
+    }
+
+    // Mark the entry COMPLETED, and select the best best entry if needed.
+    ChangeAndReselect(it, State::COMPLETED);
+
+    return true;
+}
+
+void TxRequestTracker::SetTimePoint(std::chrono::microseconds now)
+{
+    // Iterate over all CANDIDATE_DELAYED and REQUESTED from old to new, as long as they're in the past,
+    // and convert them to CANDIDATE_READY and COMPLETED respectively.
+    while (!m_index.empty()) {
+        auto it = m_index.get<ByTime>().begin();
+        if (it->GetState() == State::CANDIDATE_DELAYED && it->m_time <= now) {
+            PromoteCandidateNew(m_index.project<ByTxid>(it));
+        } else if (it->GetState() == State::REQUESTED && it->m_time <= now) {
+            MakeCompleted(m_index.project<ByTxid>(it));
+        } else {
+            break;
+        }
+    }
+
+    while (!m_index.empty()) {
+        // If time went backwards, we may need to demote CANDIDATE_BEST and CANDIDATE_READY entries back
```

Comment:
Time going backward wouldn't be a hint of some clock manipulation and thus we should interpret this as a system failure and halting ? AFAICT, `GetAdjustedTime()` can go backward due to relying on peers clocks but we are relying on `GetTime<T>` here.

Reply:
Time jumping backwards can happen for completely innocuous reasons, like leap seconds. Only monotonic clocks are guaranteed not to do that.